src/proposals/referendums_settings.cpp

```diff
@@ -0,0 +1,393 @@
+  referendums::proposal_auxiliary_tables propaux_t(contract_name, contract_name.value);
+
+  auto ritr = proposals_t.require_find(proposal_id, "referendum not found");
+  auto raitr = propaux_t.require_find(proposal_id, "refaux entry not found");
+
+  check(ritr->stage == ProposalsCommon::stage_staged, "can not update referendum, it is not staged");
+  check(ritr->type == ProposalsCommon::type_ref_setting, "referendum has to be of type settings");
+
+  uint64_t min_test_cycles = this->m_contract.config_get("refmintest"_n);
+  uint64_t test_cycles = std::get<uint64_t>(args["test_cycles"]);
+  check(test_cycles >= min_test_cycles, "the number of test cycles must be at least " + std::to_string(min_test_cycles));
+
+  uint64_t min_eval_cycles = this->m_contract.config_get("refmineval"_n);
+  uint64_t eval_cycles = std::get<uint64_t>(args["eval_cycles"]);
+  check(eval_cycles >= min_eval_cycles, "the number of eval cycles must be at least " + std::to_string(min_eval_cycles));
+
+  proposals_t.modify(ritr, contract_name, [&](auto & item) {
+    item.title = std::get<string>(args["title"]);
+    item.summary = std::get<string>(args["summary"]);
+    item.description = std::get<string>(args["description"]);
+    item.image = std::get<string>(args["image"]);
+    item.url = std::get<string>(args["url"]);
+  });
+
+  propaux_t.modify(raitr, contract_name, [&](auto & item){
+    item.special_attributes.at("setting_name") = setting_name;
+    item.special_attributes.at("is_float") = s_info->is_float;
+    if (s_info->is_float) {
+      item.special_attributes.at("new_value") = std::get<double>(args["new_value"]);
+      item.special_attributes.at("previous_value") = s_info->previous_value_double;
+    } else {
+      item.special_attributes.at("new_value") = std::get<uint64_t>(args["new_value"]);
+      item.special_attributes.at("previous_value") = s_info->previous_value_uint;
+    }
+    item.special_attributes.at("cycles_per_status") = "1," + std::to_string(test_cycles) + "," + std::to_string(eval_cycles);
+  });
+
+}
+
+void ReferendumSettings::cancel (std::map<std::string, VariantValue> & args) {
+
+  name contract_name = this->m_contract.get_self();
+  uint64_t proposal_id = std::get<uint64_t>(args["proposal_id"]);
+
+  referendums::proposal_tables proposals_t(contract_name, contract_name.value);
+  referendums::proposal_auxiliary_tables propaux_t(contract_name, contract_name.value);
+
+  auto ritr = proposals_t.require_find(proposal_id, "referendum not found");
+  auto raitr = propaux_t.require_find(proposal_id, "refaux entry not found");
+
+  check(ritr->stage == ProposalsCommon::stage_staged, "can not cancel referendum, it is not staged");
+  check(ritr->type == ProposalsCommon::type_ref_setting, "referendum has to be of type settings");
+
+  proposals_t.erase(ritr);
+  propaux_t.erase(raitr);
+
+}
+
+void ReferendumSettings::evaluate (std::map<std::string, VariantValue> & args) {
+
```

Comment:
what is the reason to pass an args object in here, vs. just the id?

Reply:
There is no difference for the moment, but I in the proposals contract for the evaluation we pass the cycle as well and as each proposal will implement it's own evaluate method I didn't want to limit the Parent class, but I can change it